File: apps/go-analytics-client/metrics.go
package main

import (
	"time"
)

// QueryExecutionMetrics represents metrics for a single query execution
type QueryExecutionMetrics struct {
	StartTime           int64   `json:"start_time"`
	EndTime             int64   `json:"end_time"`
	Success             bool    `json:"success"`
	ErrorMessage        string  `json:"error_message,omitempty"`
	RowCount            int     `json:"row_count"`
	SDKType             string  `json:"sdk_type"`
	QueryName           string  `json:"query_name"`
	DurationNanos       int64   `json:"duration_nanos"`
	DurationMs          float64 `json:"duration_ms"`
	AbsoluteStartTimeMs int64   `json:"absolute_start_time_ms"`
	AbsoluteEndTimeMs   int64   `json:"absolute_end_time_ms"`
	SequenceNumber      int     `json:"sequence_number"`
	Timestamp           int64   `json:"timestamp"`
}

// NewQueryExecutionMetrics creates a new metrics instance
func NewQueryExecutionMetrics(
	startTime, endTime time.Time,
	success bool,
	errorMessage string,
	rowCount int,
	sdkType, queryName string,
	sequenceNumber int,
	absoluteStartTimeMs int64,
) *QueryExecutionMetrics {
	durationNanos := endTime.Sub(startTime).Nanoseconds()
	absoluteEndTimeMs := absoluteStartTimeMs + (durationNanos / 1_000_000)
	
	return &QueryExecutionMetrics{
		StartTime:           startTime.UnixNano(),
		EndTime:             endTime.UnixNano(),
		Success:             success,
		ErrorMessage:        errorMessage,
		RowCount:            rowCount,
		SDKType:             sdkType,
		QueryName:           queryName,
		DurationNanos:       durationNanos,
		DurationMs:          float64(durationNanos) / 1_000_000.0,
		AbsoluteStartTimeMs: absoluteStartTimeMs,
		AbsoluteEndTimeMs:   absoluteEndTimeMs,
		SequenceNumber:      sequenceNumber,
		Timestamp:           absoluteStartTimeMs,
	}
} 

File: apps/python-analytics-client/metrics.py
"""
Query Execution Metrics

Data structure for tracking query performance metrics.
"""

import time
from dataclasses import dataclass
from typing import Optional


@dataclass
class QueryExecutionMetrics:
    """Metrics for a single query execution"""
    
    start_time: float
    end_time: float
    success: bool
    error_message: str
    row_count: int
    sdk_type: str
    query_name: str
    sequence_number: int
    absolute_start_time_ms: int
    
    def __post_init__(self):
        """Calculate derived metrics after initialization"""
        self.duration_nanos = int((self.end_time - self.start_time) * 1_000_000_000)
        self.duration_ms = (self.end_time - self.start_time) * 1000.0
        self.absolute_end_time_ms = self.absolute_start_time_ms + int(self.duration_ms)
        self.timestamp = self.absolute_start_time_ms
        
        # Convert to nanoseconds since epoch for compatibility with Go/Java
        self.start_time_ns = int(self.start_time * 1_000_000_000)
        self.end_time_ns = int(self.end_time * 1_000_000_000)
    
    def to_dict(self) -> dict:
        """Convert to dictionary for JSON serialization"""
        return {
            "start_time": self.start_time_ns,
            "end_time": self.end_time_ns,
            "success": self.success,
            "error_message": self.error_message if self.error_message else "",
            "row_count": self.row_count,
            "sdk_type": self.sdk_type,
            "query_name": self.query_name,
            "duration_nanos": self.duration_nanos,
            "duration_ms": self.duration_ms,
            "absolute_start_time_ms": self.absolute_start_time_ms,
            "absolute_end_time_ms": self.absolute_end_time_ms,
            "sequence_number": self.sequence_number,
            "timestamp": self.timestamp
        } 